### PR TITLE
General improvements to securedrop-export

### LINKED ISSUE #15: Export architecture improvement spike
For the 9/19-10/9 sprint, we've agreed to a time-boxed, 4 hour improvement spike to the current export tooling (I've increased the estimate to 8 total, to account for review and integration). The current architecture was a first pass, and a targeted improvement spike will likely identify several areas where it should be refactored or reorganized to increase maintainability and resilience.

This ticket exists to log initial observations, and PRs without corresponding issues can be tracked against it.

### PR DESCRIPTION
**Pull request by emkll**
### Description
Closes #15 
Provides improvements:
- wrap check_call to ensure exceptions are more reliably caught
- use enum to specify error codes
- fix some tests/regressions previously introduced (by decoupling tests for `set_extracted_device_name()` and `check_luks()`)

### Test plan

- [ ] Export works in Qubes using the client:
- build package on this branch
- install to sd-export-usb-template (or sd-export-usb when it's running)
- test export functionality

### CODE REVIEW

File: securedrop_export/export.py [hunk 1]
```diff
@@ -23,6 +25,41 @@
...
+
+    # General errors
+    ERROR_FILE_NOT_FOUND = 'ERROR_FILE_NOT_FOUND'
+    ERROR_EXTRACTION = 'ERROR_EXTRACTION'
+    ERROR_METADATA_PARSING = 'ERROR_METADATA_PARSING'
+    ERROR_ARCHIVE_METADATA = 'ERROR_ARCHIVE_METADATA'
+    ERROR_USB_CONFIGURATION = 'ERROR_USB_CONFIGURATION'
+    ERROR_GENERIC = 'ERROR_GENERIC'
+
+    # USB preflight related errors
```
Comment (creviera) [inline securedrop_export/export.py]: we talked already about this but it'll be a readability improvement, after this change, to move lsblk check into the usb disk preflight check and create a new printer preflight check and to use status codes:
`ERROR_DISK_NOT_FOUND`
`ERROR_DISK_PREFLIGHT_CHECK`
`ERROR_PRINTER_NOT_FOUND`
`ERROR_PRINTER_PREFLIGHT_CHECK`

File: securedrop_export/export.py [hunk 2]
```diff
@@ -23,6 +25,41 @@
 logger = logging.getLogger(__name__)
 
 
+class ExportStatus(Enum):
```
Comment (creviera) [inline securedrop_export/export.py]: :+1:

File: securedrop_export/export.py [hunk 3]
```diff
@@ -100,18 +137,15 @@ def __init__(self, archive, config_path):
         )
         self.tmpdir = tempfile.mkdtemp()
 
+    def safe_check_call(self, command, error_message):
```
Comment (creviera) [inline securedrop_export/export.py]: this make me so happy, later we could even add type hints

File: securedrop_export/export.py [hunk 4]
```diff
@@ -120,48 +154,43 @@ def exit_gracefully(self, msg, e=False):
...
         solutions for mimetype handling, which we want to avoid.
         """
-        sys.stderr.write(msg)
-        sys.stderr.write("\n")
         logger.info('Exiting with message: {}'.format(msg))
-        if e:
+        if not e:
+            sys.stderr.write(msg)
+            sys.stderr.write("\n")
+        else:
```
Comment (creviera) [inline securedrop_export/export.py]: :+1:  this should fix the bug where we were seeing extra output in our error status code (will test after code review)

File: securedrop_export/export.py [hunk 5]
```diff
@@ -173,7 +202,7 @@ def check_usb_connected(self):
                 stderr=subprocess.PIPE)
             self.exit_gracefully("USB_CONNECTED")
         except subprocess.CalledProcessError:
-            self.exit_gracefully("USB_NOT_CONNECTED")
+            self.exit_gracefully(ExportStatus.USB_NOT_CONNECTED.value)
```
Comment (creviera) [inline securedrop_export/export.py]: why not use `safe_check_call` here?

File: securedrop_export/export.py [hunk 6]
```diff
@@ -228,33 +257,33 @@ def unlock_luks_volume(self, encryption_key):
...
+        )
+        self.safe_check_call(
+            command=["sudo", "chown", "-R", "user:user", self.mountpoint],
+            error_message=ExportStatus.ERROR_USB_MOUNT.value
+        )
 
     def copy_submission(self):
         # move files to drive (overwrites files with same filename) and unmount drive
+        # we don't use safe_check_call here because we must lock and
+        # unmount the drive as part of the finally block
```
Comment (creviera) [inline securedrop_export/export.py]: :+1:

File: tests/test_export.py [hunk 7]
```diff
@@ -236,37 +208,63 @@ def test_usb_precheck_connected(mocked_call, capsys):
...
+    mocked_exit = mock.patch("export.exit_gracefully", return_value=0)
+
+    with pytest.raises(SystemExit) as sysexit:
+        submission.check_luks_volume()
+        mocked_exit.assert_called_once_with(expected_message)
+    assert sysexit.value.code == 0
+
+
 @mock.patch("subprocess.check_output", return_value=SAMPLE_OUTPUT_ONE_PART)
-def test_luks_precheck_encrypted(mocked_call, capsys):
```
Comment (creviera) [inline tests/test_export.py]: i like this test name :smirk:

File: securedrop_export/export.py (hunk 5 above)
Comment (emkll) [reply securedrop_export/export.py]: if you are referring to https://github.com/freedomofpress/securedrop-export/pull/28/files#diff-f7920fa15cd44ca4817dafa773cf5372R200 specifically, the reason is that is `check_output` and we would need to create a new wrapper method for it. Given it was used only once, it was not worth the maintenance cost. What do you think?

Comment (creviera) [reply securedrop_export/export.py]: oh i see, the wrapper is for `check_call` not `check_output`. looks like we have another call to `check_output` when getting the printer uri so a new wrapper for `check_output` might be a good idea, but I'm fine with delaying until after we finish printer support in the client.

**Pull request merged**